### PR TITLE
ImportC: add semantic() for compound literals

### PR DESCRIPTION
**Pull request by WalterBright**
Also added hdrgen for compound literals so they can be pretty-printed.

The semantic routine relies on initializers getting better at being converted to an Expression.

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -3238,6 +3238,24 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+        static if (LOGSEMANTIC)
+        {
+            printf("CompoundLiteralExp::semantic('%s')\n", cle.toChars());
+        }
+        Type t = cle.type.typeSemantic(cle.loc, sc);
+        auto init = initializerSemantic(cle.initializer, sc, t, INITnointerpret);
+        auto e = initializerToExpression(init, t);
+        if (!e)
+        {
+            error(cle.loc, "cannot convert initializer `%s` to expression", init.toChars());
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d:3252]: This error is not covered by tests.

File: src/dmd/hdrgen.d [hunk 2]
```diff
@@ -2044,6 +2044,14 @@ public:
         buf.writeByte(')');
     }
 
+    override void visit(CompoundLiteralExp e)
+    {
+        buf.writeByte('(');
+        typeToBuffer(e.type, null, buf, hgs);
+        buf.writeByte(')');
+        e.initializer.initializerToBuffer(buf, hgs);
```
Comment (RazvanN7) [inline src/dmd/hdrgen.d:2052]: This code is not covered by tests.

File: test/runnable/cstuff2.c [hunk 3]
```diff
@@ -188,6 +188,15 @@ void test11()
 
 /*********************************/
 
+void test12()
+{
+    int i;
+    i = (int) { 3 };
+    if (i != 3) { printf("test12\n"); exit(1); }
```
Comment (RazvanN7) [inline test/runnable/cstuff2.c:195]: This should be more thoroughly tested:

- empty type
- empty initializer compound statement
- more than 1 expressions inside the compound statement
- the hdrgen code should also be tested

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/expressionsem.d:3252]: Earlier, I mentioned that I was not going to do coverage for error messages, due to the fluidity and changes going on. Instead, I concentrate on getting correct examples to work, as those shouldn't change.

File: src/dmd/hdrgen.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/hdrgen.d:2052]: Right, because the only way to test them is with error messages.

File: test/runnable/cstuff2.c (hunk 3 above)
Comment (WalterBright) [reply test/runnable/cstuff2.c:195]: There are still many shortcomings in the initializer semantics that I need to address. But that isn't what this PR is about.

**Review by PetarKirov — APPROVED** [review]

**Comment by PetarKirov** [comment]
Given that @WalterBright prefers to improve / fix stuff incrementally regarding ImportC and there will soon be more PRs that depend on this one, I will go ahead and merge it.

**Pull request merged**